dockerize and switch to SMTP to send email

leonardow-unep-wcmc opened the pull request:
- Tried nodejs 12, but it doesn't work with current version node-sass. Stay at nodejs v10.
- Tried ruby 2.6, but it doesn't work with current version bcrypt. Stay at ruby 2.5.
- switch to SMTP (sendgrid) approach to send email, two new env are needed: MAIL_USERNAME, MAIL_PASSWORD
- dockerize for local development. Need .env.development with SHIPMENTS_API_TOKEN, value from staging.
- Dockerfile for staging/production.
- New ENV for database config (for kamal deploy):
  - DATABASE_USERNAME
  - DATABASE_PASSWORD
  - DATABASE_PORT
  - DATABASE_HOST

leonardow-unep-wcmc committed:
dockerize and switch to SMTP to send email

leonardow-unep-wcmc committed:
fix ruby version

leonardow-unep-wcmc committed:
production dockerfile

leonardow-unep-wcmc committed:
production dockerfile

pdl commented on Dockerfile:
```diff
@@ -0,0 +1,71 @@
+
+# Precompiling assets for production
+# TODO: need to change when upgrade Rails.
+RUN SECRET_KEY_BASE=dummy ./bin/rails assets:precompile
```
Is `SECRET_KEY_BASE` a placeholder for future rails here? Or is it supposed to be set with a `.env.staging` file - in which case doesn't it get overridden by `dummy` here?

leonardow-unep-wcmc replied:
SECRET_KEY_BASE is essential in staging and production in Rails, it use to encrypt the cookie/session and password (but not devise).

In the old days, cap deploy is basiclly SSH into the server and run the assets precompile which the .env is there, all good, but actually not really using the it (assets don't need that secret at all).

When we use docker nowadays, we build the image without any secret, but due to the rails stack it still need that to make it run. Newer rails come with a new env for that in [Dockerfile](https://railsdiff.org/7.0.8.7/7.1.5.1#diff-466a28a0e93935ce250159682062e5a94698a3d8):
```
RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile
```
But it is not available in the rails version using in this project.

In my knowledge `RUN SECRET_KEY_BASE=dummy ./bin/rails assets:precompile` inline environment variable only available in that line, not retain in other docker command when build, as well as the container.

There is one SECRET_KEY_BASE in .env in staging and production server. And later if we deploy using kamal, we need to set that in kamal config too. You may find some newer project do not need that env because the value is stored in Rails credential. However this project still using secrets.yml with ENV, thats why.

pdl replied:
Oh it is me that is the dummy, I got it confused with RAILS_MASTER_KEY. Yep, I agree this is safe.

leonardow-unep-wcmc replied:
One more thing we need to keep it in mind about assets.

In a perfect setup, the assets should upload to CDN instead of hosting locally.
If we host locally, it should put in a share folder (cap shared/kamal host volume), so that the out-dated assets (from previous releases) still accessable by the end-user.

All assets is build with a checksum in the file name, to ensure the user always loading the latest assets.
However in some scenario, end-user still need to access some old/out-dated assets. For example, if we use an image in the email. The email content point to the old image which may no longer avaiable in the latest release. If we not setup the assets folder properly, that image in the email will be broken.

Good news is, seems all our email in WCMC is simple text, so we don't need to worry about this.

leonardow-unep-wcmc commented on .node-version:
```diff
@@ -0,0 +1 @@
+10.24.1
```
The readme.md has mention it. It is for webpacker i believe.

leonardow-unep-wcmc committed:
rename Dockerfile, follow latest DevOps practice

leonardow-unep-wcmc committed:
up page for healthcheck

leonardow-unep-wcmc committed:
upgrade ruby to 2.6 (ruby 2.5 image too old, apt no longer work)

leonardow-unep-wcmc committed:
upgrade bcrypt to make it works with Ruby 2.6

pdl committed:
chore: sapi

pdl commented:
OK, I've tried getting it working so we can use local SAPI as the data source: needed to make a change to SAPI as well https://github.com/unepwcmc/SAPI/commit/def1625a7731ee32c505017ebb63e0bb594201c9

Note, however, that unless the following ports are all set (and set differently between this tool and SAPI), then trying to run both at once fails because of host port conflict.

```
CONTAINER_MAILCATCHER_PORT
CONTAINER_DB_PORT
CONTAINER_RAILS_PORT
```

pdl commented:
I'm happy this works locally.

pdl committed:
fix: respect SPECIES_API_URL env var

leonardow-unep-wcmc commented on compose.yaml:
```diff
@@ -0,0 +1,61 @@
+    environment:
+      DATABASE_HOST: cites-compliance-tool-db
+      # SPECIES_API_URL: "${SPECIES_API_URL:-https://shipments.sapi-staging.linode.unep-wcmc.org/api/v1}"
+      SPECIES_API_URL: "${SPECIES_API_URL:-http://sapi-rails:3000/api/v1}"
```
@pdl I would say the default should be staging, so this project out of the box working. Can be overrided by using .env file in the project root.

leonardow-unep-wcmc committed:
kamal 2 config

leonardow-unep-wcmc committed:
enhance Dockerfile for local development

leonardow-unep-wcmc committed:
Dockerfile for production, tested locally

leonardow-unep-wcmc commented:
## Changes
- **Dockerfile**: Need hardcore `--platform=linux/amd64` for nodejs installation.
- **Ruby**: At the end able to upgraded to `2.6.10`.  
- **database.yml**: Updated to use environment variables. (Replaced by a copy in `staging/production` via the Capistrano deploy.)  
- **deploy/**: Added a folder containing its own `Gemfile` and `.ruby-version` to pin the Ruby and Kamal versions for deployment.  
- **Healthcheck**: Introduced a `/up` endpoint.  
- **Kamal configs**: Added Kamal configuration files.  
- **Kamal pre-build hook**: Enabled the `pre-build` hook in Kamal.

## Notes
- Run `docker compose up -d` to start the local development environment.  
- RVM and NVM are no longer used for Ruby/Node installation; they were only required for Capistrano deployments.  
- `Dockerfile.deploy` has been tested locally in production mode.  
- Environments secrets such as DB credentials, have been extracted from the existing server and added to the project’s GitHub Environments.  
- Assets were previously served by Nginx; they’re now served by Puma. The `RAILS_SERVE_STATIC_FILES` environment variable has been added in `config/deploy.yml`.  
- The `RAILS_LOG_TO_STDOUT` environment variable has also been added in `config/deploy.yml` so that logs are written to stdout. 

## TODO
- Remove Capistrano (not done yet).  
- Finalise staging and production server addresses (TBC).

pdl commented:
Looks good to me. Needs DevOps involvement before we merge.

leonardow-unep-wcmc committed:
Update Kamal deploy configuration
